File: src/main/java/com/leijendary/spring/boot/template/error/AccessDeniedExceptionHandler.java
package com.leijendary.spring.boot.template.error;

import static com.leijendary.spring.boot.template.util.RequestContext.getLocale;
import static org.springframework.http.HttpStatus.FORBIDDEN;

import com.leijendary.spring.boot.template.data.response.ErrorResponse;

import org.springframework.context.MessageSource;
import org.springframework.core.annotation.Order;
import org.springframework.security.access.AccessDeniedException;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.ResponseStatus;
import org.springframework.web.bind.annotation.RestControllerAdvice;

import lombok.RequiredArgsConstructor;

@RestControllerAdvice
@Order(1)
@RequiredArgsConstructor
public class AccessDeniedExceptionHandler {

    private final MessageSource messageSource;

    @ExceptionHandler(AccessDeniedException.class)
    @ResponseStatus(FORBIDDEN)
    public ErrorResponse catchAccessDenied(final AccessDeniedException exception) {
        final var code = "access.denied";
        final var arguments = new Object[0];
        final var message = messageSource.getMessage(code, arguments, getLocale());

        return ErrorResponse.builder()
                .addError("Authorization", code, message + ": " + exception.getMessage())
                .status(FORBIDDEN)
                .build();
    }
}

File: src/main/kotlin/com/leijendary/spring/template/core/error/AccessDeniedExceptionHandler.kt
package com.leijendary.spring.template.core.error

import com.leijendary.spring.template.core.model.ErrorModel
import com.leijendary.spring.template.core.util.RequestContext
import org.springframework.context.MessageSource
import org.springframework.core.annotation.Order
import org.springframework.http.HttpStatus.FORBIDDEN
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.ResponseStatus
import org.springframework.web.bind.annotation.RestControllerAdvice

@RestControllerAdvice
@Order(1)
class AccessDeniedExceptionHandler(private val messageSource: MessageSource) {
    @ExceptionHandler(AccessDeniedException::class)
    @ResponseStatus(FORBIDDEN)
    fun catchAccessDenied(exception: AccessDeniedException): List<ErrorModel> {
        val code = "access.denied"
        val arguments = arrayOfNulls<Any>(0)
        val message = messageSource.getMessage(code, arguments, RequestContext.locale)
        val error = ErrorModel(
            mutableListOf("header", "authorization"),
            code,
            message + ": " + exception.message
        )

        return listOf(error)
    }
}
